kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java

```diff
@@ -20,104 +35,319 @@
 
+    log.debug("No proto schema found, skipping buf for topic {}", msg.topic());
+
     return this.schemaRegistryAwareRecordSerDe.deserialize(msg);
   }
 
-  @Nullable ProtoSchema protoSchemaFromHeaders(Headers headers) {
-    // extract PROTOBUF_TYPE header
+  private DeserializedKeyValue deserializeProto(ConsumerRecord<Bytes, Bytes> msg,
+        String keyFullyQualifiedType,
+        String valueFullyQualifiedType) {
+    Descriptor keyDescriptor = null;
+    if (keyFullyQualifiedType != null) {
+      keyDescriptor = getDescriptor(keyFullyQualifiedType);
+      if (keyDescriptor == null) {
+        log.warn("No key descriptor found for topic {} with schema {}", msg.topic(), keyFullyQualifiedType);
+      }
+    }
+
+    Descriptor valueDescriptor = null;
+    if (valueFullyQualifiedType != null) {
+      valueDescriptor = getDescriptor(valueFullyQualifiedType);
+      if (valueDescriptor == null) {
+        log.warn("No value descriptor found for topic {} with schema {}", msg.topic(), valueFullyQualifiedType);
+      }
+    }
+
+    return this.deserializeProtobuf(msg, keyDescriptor, valueDescriptor);
+  }
+
+  @Nullable
+  ProtoSchema protoKeySchemaFromHeaders(Headers headers) {
+    // Get PROTOBUF_TYPE_KEY header.
     String fullyQualifiedTypeName = null;
-    for (Header header : headers.headers("PROTOBUF_TYPE")) {
+    for (Header header : headers.headers("PROTOBUF_TYPE_KEY")) {
       fullyQualifiedTypeName = new String(header.value());
     }
 
     if (fullyQualifiedTypeName == null) {
       return null;
     }
 
-    // extract PROTOBUF_SCHEMA_ID header
-    String schemaID = null;
-    for (Header header : headers.headers("PROTOBUF_SCHEMA_ID")) {
-      schemaID = new String(header.value());
+    ProtoSchema ret = new ProtoSchema();
+    ret.setFullyQualifiedTypeName(fullyQualifiedTypeName);
+    return ret;
+  }
+
+  @Nullable
+  ProtoSchema protoValueSchemaFromHeaders(Headers headers) {
+    // Get PROTOBUF_TYPE_VALUE header.
+    String fullyQualifiedTypeName = null;
+    for (Header header : headers.headers("PROTOBUF_TYPE_VALUE")) {
+      fullyQualifiedTypeName = new String(header.value());
+    }
+
+    if (fullyQualifiedTypeName == null) {
+      return null;
     }
 
     ProtoSchema ret = new ProtoSchema();
     ret.setFullyQualifiedTypeName(fullyQualifiedTypeName);
-    ret.setSchemaID(schemaID);
     return ret;
   }
 
-  @Nullable ProtoSchema protoSchemaFromTopic(String topic) {
+  @Nullable
+  ProtoSchema protoSchemaFromTopic(String topic) {
     if (!topic.contains(".proto.")) {
       return null;
     }
 
-    // extract fqtn
+    // Extract the fully qualified type name from the topic.
+    // E.g., my-topic.proto.service.v1.User -> service.v1.User.
     String[] parts = topic.split("\\.proto\\.");
     ProtoSchema ret = new ProtoSchema();
     ret.setFullyQualifiedTypeName(parts[parts.length - 1]);
     return ret;
   }
 
+  private DeserializedKeyValue deserializeProtobuf(ConsumerRecord<Bytes, Bytes> msg,
```

Comment:
any time you do `new String(someBytes)` and you don't specify the encoding, you are at the mercy of your OS default encoding. Use `new String(someBytes, StandardCharsets.UTF_8)`